RecoCTPPS/TotemRPLocal/src/TotemRPLocalTrackFitterAlgorithm.cc

```diff
@@ -0,0 +1,203 @@
+  rp_topology_stripaxis_clhep.setY(rp_topology_stripaxis.y());
+  rp_topology_stripaxis_clhep.setZ(rp_topology_stripaxis.z());
+ 
+  TVector3 rd_dir = convert3vector( tot_rp_geom.LocalToGlobalDirection(det_id, rp_topology_stripaxis_clhep ) );
+  
+  TVector2 v(rd_dir.X(), rd_dir.Y());
+  det_algebra_obj.readout_direction_ = v.Unit();
+  det_algebra_obj.rec_u_0_ = 0.0;
+  det_algebra_obj.available_ = true;
+  det_algebra_obj.rec_u_0_ = - (det_algebra_obj.readout_direction_ * det_algebra_obj.centre_of_det_global_position_.XYvector());
+  
+  return det_algebra_obj;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+TotemRPLocalTrackFitterAlgorithm::RPDetCoordinateAlgebraObjs*
+TotemRPLocalTrackFitterAlgorithm::GetDetAlgebraData(unsigned int det_id, const TotemRPGeometry& tot_rp_geom)
+{
+  auto it = det_data_map_.find(det_id);
+  if (it != det_data_map_.end())
+  {
+    return &(it->second);
+  } else {
+    det_data_map_[det_id] = PrepareReconstAlgebraData(det_id, tot_rp_geom);
+    return &det_data_map_[det_id];
+  }
+}
+
+//----------------------------------------------------------------------------------------------------
+
+bool TotemRPLocalTrackFitterAlgorithm::FitTrack(const edm::DetSetVector<TotemRPRecHit> &hits, double z_0,
+    const TotemRPGeometry &tot_geom, TotemRPLocalTrack &fitted_track)
+{
+  fitted_track.setValid(false);
+
+  // bind hits with their algebra objects
+  struct HitWithAlg
+  {
+    unsigned int detId;
+    const TotemRPRecHit *hit;
+    RPDetCoordinateAlgebraObjs *alg;
+  };
+  
+  vector<HitWithAlg> applicable_hits;
+
+  for (auto &ds : hits)
+  {
+    unsigned int detId = ds.detId();
+
+    for (auto &h : ds)
+    {
+      RPDetCoordinateAlgebraObjs * alg = GetDetAlgebraData(detId, tot_geom);
+      if (alg->available_)
+        applicable_hits.push_back({ detId, &h, alg});
+    }
+  }
+  
+  if (applicable_hits.size() < 5)
+    return false;
+  
+  TMatrixD H(applicable_hits.size(), 4);
+  TVectorD V(applicable_hits.size());
+  TVectorD V_inv(applicable_hits.size());
+  TVectorD U(applicable_hits.size());
+  
+  for(unsigned int i = 0;  i < applicable_hits.size(); ++i)
+  {
+    RPDetCoordinateAlgebraObjs *alg_obj = applicable_hits[i].alg;
+
+    H(i,0) = alg_obj->readout_direction_.X();
+    H(i,1) = alg_obj->readout_direction_.Y();
+    double delta_z = alg_obj->centre_of_det_global_position_.Z()-z_0;
+    H(i,2) = alg_obj->readout_direction_.X()*delta_z;
+    H(i,3) = alg_obj->readout_direction_.Y()*delta_z;
+    double var = applicable_hits[i].hit->getSigma();
+    var*=var;
+    V[i] = var;
+    V_inv[i] = 1.0/var;
+    U[i] = applicable_hits[i].hit->getPosition() - alg_obj->rec_u_0_;
+  }
+
+  TMatrixD H_T_V_inv(TMatrixD::kTransposed, H);
+  MultiplyByDiagonalInPlace(H_T_V_inv, V_inv);
+  TMatrixD V_a(H_T_V_inv);
+  TMatrixD V_a_mult(V_a, TMatrixD::kMult, H);
+  try
+  {
+    V_a_mult.Invert();
+  }
+  catch (cms::Exception &e)
```

Comment:
I simulated what happens if the matrix is really singular:
```
An exception of category 'FatalRootError' occurred while
   [0] Processing run: 273062 lumi: 1 event: 2
   [1] Running path 'p'
   [2] Calling event method for module TotemRPLocalTrackFitter/'totemRPLocalTrackFitter'
   Additional Info:
      [a] Fatal Root Error: @SUB=TDecompLU::DecomposeLUCrout
matrix is singular
```
and cmsRun stops. The message is not really descriptive and definitely we don't want the processing to stop because of weird numerical conditions in one track fit (very rare as I said earlier). Thus I would say that the `catch` is not a bad solution.